plugin/trino-snowflake/src/main/java/io/trino/plugin/snowflake/SnowflakeClient.java

```diff
@@ -0,0 +1,565 @@
+        return ColumnMapping.longMapping(
+                TimeType.createTimeType(precision),
+                (resultSet, columnIndex) -> {
+                    LocalTime time = SNOWFLAKE_TIME_FORMATTER.parse(resultSet.getString(columnIndex), LocalTime::from);
+                    return Timestamps.round(time.toNanoOfDay() * PICOSECONDS_PER_NANOSECOND, 12 - precision);
+                },
+                timeWriteFunction(precision),
+                PredicatePushdownController.FULL_PUSHDOWN);
+    }
+
+    private static LongWriteFunction snowFlaketimeWriter(Type type)
+    {
+        return timeWriteFunction(((TimeType) type).getPrecision());
+    }
+
+    private static LongWriteFunction timeWriteFunction(int precision)
+    {
+        checkArgument(precision <= SNOWFLAKE_MAX_SUPPORTED_TIMESTAMP_PRECISION, "Unsupported precision: %s", precision);
+        String bindExpression = format("CAST(? AS time(%s))", precision);
+        return new LongWriteFunction()
+        {
+            @Override
+            public String getBindExpression()
+            {
+                return bindExpression;
+            }
+
+            @Override
+            public void set(PreparedStatement statement, int index, long picosOfDay)
+                    throws SQLException
+            {
+                picosOfDay = Timestamps.round(picosOfDay, 12 - precision);
+                if (picosOfDay == Timestamps.PICOSECONDS_PER_DAY) {
+                    picosOfDay = 0;
+                }
+                LocalTime localTime = LocalTime.ofNanoOfDay(picosOfDay / PICOSECONDS_PER_NANOSECOND);
+                // statement.setObject(.., localTime) would yield incorrect end result for 23:59:59.999000
+                statement.setString(index, SNOWFLAKE_TIME_FORMATTER.format(localTime));
+            }
+        };
+    }
+
+    private static ColumnMapping timestampTzColumnMapping(JdbcTypeHandle typeHandle)
+    {
+        int precision = typeHandle.getRequiredDecimalDigits();
+        String jdbcTypeName = typeHandle.getJdbcTypeName()
+                .orElseThrow(() -> new TrinoException(JDBC_ERROR, "Type name is missing: " + typeHandle));
+        int type = typeHandle.getJdbcType();
+        log.debug("timestampTZColumnMapping: jdbcTypeName(%s):%s precision:%s", type, jdbcTypeName, precision);
+
+        if (precision <= 3) {
+            return ColumnMapping.longMapping(
+                    createTimestampWithTimeZoneType(precision),
+                    (resultSet, columnIndex) -> {
+                        ZonedDateTime timestamp = SNOWFLAKE_DATETIME_FORMATTER.parse(resultSet.getString(columnIndex), ZonedDateTime::from);
+                        return DateTimeEncoding.packDateTimeWithZone(timestamp.toInstant().toEpochMilli(), timestamp.getZone().getId());
+                    },
+                    timestampWithTZWriter(),
```

Comment:
convention is to call these something like `typeWriteFunction` so something like `timestampWithTimezoneWriteFunction`

(make it easier for other people familiar with other connectors to follow the code)